samples/bluetooth/central_peripheral_hr/src/main.c

```diff
@@ -0,0 +1,458 @@
+			param = BT_LE_CONN_PARAM_DEFAULT;
+			create_param = BT_CONN_LE_CREATE_CONN;
+			create_param->options |= BT_CONN_LE_OPT_CODED;
+			err = bt_conn_le_create(addr, create_param, param,
+						&conn_central);
+			if (err) {
+				printk("Create connection with Coded PHY support failed (err %d)\n",
+				       err);
+
+				printk("Creating non-Coded PHY connection\n");
+				create_param->options &= ~BT_CONN_LE_OPT_CODED;
+				err = bt_conn_le_create(addr, create_param,
+							param, &conn_central);
+				if (err) {
+					printk("Create connection failed (err %d)\n", err);
+					start_scan();
+				}
+			}
+
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
+			 struct net_buf_simple *ad)
+{
+	char dev[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(addr, dev, sizeof(dev));
+	printk("[DEVICE]: %s, AD evt type %u, AD data len %u, RSSI %i\n",
+	       dev, type, ad->len, rssi);
+
+	/* We're only interested in legacy connectable events or
+	 * possible extended advertising that are connectable.
+	 */
+	if (type == BT_GAP_ADV_TYPE_ADV_IND ||
+	    type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND ||
+	    type == BT_GAP_ADV_TYPE_EXT_ADV) {
+		bt_data_parse(ad, eir_found, (void *)addr);
+	}
+}
+
+static void start_scan(void)
+{
+	int err;
+
+	/* Use active scanning and disable duplicate filtering to handle any
+	 * devices that might update their advertising data at runtime.
+	 */
+	struct bt_le_scan_param scan_param = {
+		.type       = BT_LE_SCAN_TYPE_ACTIVE,
+		.options    = BT_LE_SCAN_OPT_CODED,
+		.interval   = BT_GAP_SCAN_FAST_INTERVAL,
+		.window     = BT_GAP_SCAN_FAST_WINDOW,
+	};
+
+	err = bt_le_scan_start(&scan_param, device_found);
+	if (err) {
+		printk("Scanning with Coded PHY support failed (err %d)\n", err);
+
+		printk("Scanning without Coded PHY\n");
+		scan_param.options &= ~BT_LE_SCAN_OPT_CODED;
+		err = bt_le_scan_start(&scan_param, device_found);
+		if (err) {
+			printk("Scanning failed to start (err %d)\n", err);
+			return;
+		}
+	}
+
+	printk("Scanning successfully started\n");
+}
+
+static bool hrf_ntf_enabled;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL,
+		      BT_UUID_16_ENCODE(BT_UUID_HRS_VAL),
+		      BT_UUID_16_ENCODE(BT_UUID_BAS_VAL),
+		      BT_UUID_16_ENCODE(BT_UUID_DIS_VAL)),
+#if defined(CONFIG_BT_EXT_ADV)
+	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, sizeof(CONFIG_BT_DEVICE_NAME) - 1),
+#endif /* CONFIG_BT_EXT_ADV */
+};
+
+#if !defined(CONFIG_BT_EXT_ADV)
+static const struct bt_data sd[] = {
+	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, sizeof(CONFIG_BT_DEVICE_NAME) - 1),
+};
+#endif /* !CONFIG_BT_EXT_ADV */
+
+/* Use atomic variable, 2 bits for connection and disconnection state */
+static ATOMIC_DEFINE(state, 2U);
+
+#define STATE_CONNECTED    1U
+#define STATE_DISCONNECTED 2U
```

Comment:
Why wouldn't you use the more common pattern for this:
```
enum {
	STATE_CONNECTED,
	STATE_DISCONNECTED,

	NUM_STATE_BITS,
};

ATOMIC_DEFINE(state, NUM_STATE_BITS);
```

It also seems an oversight that your original defines start with 1 and not 0, since the `atomic_set_bit()` etc count from 0. I.e. you're actually setting the third bit when you pass `STATE_DISCONNECTED`, which happens to work even though you pass `2` to `ATOMIC_DEFINE()`, since it'll round up to the nearest word size.

Reply:
Since the states are mutually exclusive, they could also be a volatile bool or just a single atomic value, unless they can both be set at the same time?

Reply:
@Thalley yes, that's a fair point. Somehow I didn't even consider that since the code was using `atomic_set_bit()`, etc.